subsys/bluetooth/host/gatt.c

```diff
@@ -1833,6 +1833,14 @@ ssize_t bt_gatt_attr_read(struct bt_conn *conn, const struct bt_gatt_attr *attr,
 		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
 	}
 
+	if (value_len != 0U && value == NULL) {
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	if (value_len == 0U) {
+		return 0;
+	}
+
```

Comment:
This bit of code is redundant, since `len` will be zero.

It's also insufficient to prevent calling the `memcpy` with a one-past-the-object, [which the C standard forbids by how it's worded](https://stackoverflow.com/questions/29844298/is-it-legal-to-call-memcpy-with-zero-length-on-a-pointer-just-past-the-end-of-an). I'm not sure of the reason for why it forbids this. It could have something to do with pointer arithmetic at the end of address space maybe.

I think the correct thing to do here is surround the `memcpy` below in `if (len != 0)`.

Reply:
It's early termination. 

The `len = MIN(buf_len, value_len - offset);` calculation would also be redundant if `value_len == 0`. 

If you prefer I can replace the early terminator and guard the `len = MIN(buf_len, value_len - offset);` and `memcpy` with `if (value_len != 0)` so that it becomes

```c
	if (value_len != !0U) {
		len = MIN(buf_len, value_len - offset);

		memcpy(buf, (uint8_t *)value + offset, len);
	} else {
		len = value_len;
	}

	LOG_DBG("handle 0x%04x offset %u length %u", attr->handle, offset, len);
	
	return len;
```

Let me know if you prefer that

Reply:
Ended up doing that anyways, as the early termination would also omit the log statement